infrastructure/src/main/java/org/corfudb/infrastructure/management/failuredetector/DecisionMakerAgent.java

```diff
@@ -29,13 +30,13 @@ public Optional<String> findDecisionMaker() {
             return Optional.empty();
         }
 
-        Set<String> writableNodes = writableNodes();
+        Set<String> healthyNodes = healthyNodes();
 
         return clusterAdvisor
-                .findDecisionMaker(clusterState, writableNodes)
+                .findDecisionMaker(clusterState, healthyNodes)
                 .map(NodeRank::getEndpoint)
-                //filter out read-only nodes
-                .filter(writableNodes::contains)
+                //filter out unhealthy nodes
+                .filter(healthyNodes::contains)
```

Comment:
Storage related IO exceptions can happen in different places, but this code only assumes that it can happen during transaction commit. Any exceptions related to reading or corrupted checksums are not accounted for. This is why I suggest that the failure detector should operate on a generic category of transient/permanent IO errors.